### ISSUE TITLE
Optimize buyCoffee function in BuyMeACoffee.sol for gas efficiency

### ISSUE THREAD

**Issue by 2911rahulpathak** [coinbase/build-onchain-apps#195]
### Describe the solution you'd like

The current implementation of the `buyCoffee` function uses `memory` for string arguments. I propose optimizing this function for gas efficiency by changing the argument types from `memory` to `calldata` as the value of the argument variable paased is not being changed inside the function.

**Benefits**:

1. Gas Savings: Using `calldata` for string arguments can potentially result in gas savings during function execution.
2. Efficiency: The change aims to improve the overall gas efficiency of the contract.

![Screenshot (4365)](https://github.com/coinbase/build-onchain-apps/assets/73342162/f63d7566-8024-4512-93fe-b771425a8839)

## Alternatives considered

1. **Current Implementation:**
   The current implementation uses `memory` for string arguments...

2. **Calldata Approach:**
   An alternative approach is to use `calldata` instead of `memory` for string arguments...

3. **Gas Efficiency vs. Readability:**
   One consideration is the potential trade-off between gas efficiency and code readability...

4. **Gas Profiling:**
   Prior to making changes, gas profiling and benchmarking different implementations could be conducted.
   
 ## Questions for Discussion

- What are your thoughts on changing the `buyCoffee` function to use `calldata` for string arguments?
- Are there alternative solutions or considerations that should be taken into account?
- How will this change impact contract compatibility and upgrades?

**Comment by Zizzamia** [comment]
That's super dope @2911rahulpathak, great call out.

Learning this for the first time, and it's super interesting to see the difference between `memory` and `calldata` ([here a good link](https://ethereum.stackexchange.com/questions/74442/when-should-i-use-calldata-and-when-should-i-use-memory)).

And based on my understanding `memory` seems like a superset of `calldata`, as both they exist to store data during the execution of a `function`, but `memory` is mutable were `calldata` is not. So now the question is, are we interested to keep for this example those variable mutable? 

For the `buyCoffee` both `userName` and `message` seem both fine to use `calldata` instead. cc @Sneh1999 thoughts on this?

Btw, I am curios to know how folks do benchmarks between different implementation?

Btw, I found this article super interesting as well https://www.alchemy.com/overviews/solidity-gas-optimization#8-store-data-in-calldata-instead-of-memory-for-certain-function-parameters

@2911rahulpathak please go ahead and start a PR, with updated tests and docs, and would love to review it with @Sneh1999 .

If you don't have time, let me know and I will take care of it.

Thank you again, really appreciated the hint.

**Comment by 2911rahulpathak** [comment]
Hi @Zizzamia,

Thanks for your quick response and the helpful insights! I've just created a pull request for the optimization, addressing your points:

- Mutability: I've confirmed that userName and message don't require mutability within buyCoffee, making calldata suitable.

- Documentation: I've updated the documentation to reflect the change and its rationale.

Here's the link to the pull request: https://github.com/coinbase/build-onchain-apps/pull/196

Looking forward to your review!

Also, I wanted to know more about this⬇️  

![Screenshot (4367)](https://github.com/coinbase/build-onchain-apps/assets/73342162/df118fe6-be16-4c0d-a5cd-ad06c6e30392)

It shows: All checks are failed. Is this failing because of test cases? or something else?
  
Thanks again for your support!
